client/notices.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+type NotifyOptions struct {
+	// Key is the client notice's key. Must be in "domain.com/key" format.
+	Key string
+
+	// RepeatAfter, if provided, allows the notice to repeat after this duration.
+	RepeatAfter time.Duration
+
+	// Data are optional key=value pairs for this occurrence of the notice.
+	Data map[string]string
+}
+
+// Notify records an occurrence of a "client" notice with the specified options,
+// returning the notice ID.
+func (client *Client) Notify(opts *NotifyOptions) (string, error) {
+	var payload = struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after,omitempty"`
+		Data        map[string]string `json:"data,omitempty"`
+	}{
+		Action: "add",
+		Type:   "client",
+		Key:    opts.Key,
+		Data:   opts.Data,
+	}
+	if opts.RepeatAfter != 0 {
+		payload.RepeatAfter = opts.RepeatAfter.String()
+	}
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&payload); err != nil {
+		return "", err
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{}
+	_, err := client.doSync("POST", "/v1/notices", nil, nil, &body, &result)
+	if err != nil {
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
```

Comment:
This was called ```NoticeFilters``` in the state implementation. I think the reasoning is that in the client context, these are options to the client request itself. But for me as an outside person, having filter in the name is more descriptive, and having it the same name as the state implementation feels consistent.

Reply:
Per above comment, all the other `Client` methods take a `FooOptions`, so I don't think these should diverge. I agree it'd be nice to match the name in state, but users of client shouldn't need to know about state so I think that's a non-issue. I think we should go for consistency with the parameter name of the other `Client` methods here.